Add a new API

gws opened the pull request:
In preparation for future address geocoding improvements, we need to change the way we parse SmartyStreets' API responses.

The existing code has the following assumption at its core:

```clojure
(defn- query->response [responses idx query]
  (first (filter #(= idx (:input_index %)) responses)))
```

The problem with the above is that responses with multiple candidate addresses have all but the first candidate address stripped from the response. In order to change this assumption without breaking existing consumers, we need a new API that better models address responses. This new API has been added in the form of new namespaces. No code has been removed, so existing software should be able to pull down the new artifact and continue to use the library.

The overall goal of the changes is to wrap the SmartyStreets API as directly as possible. Aside from the general structure of requests and responses, we are agnostic to particular fields in requests or responses.

In general, the only transformations done to the API responses are to map flat responses containing `input_index` and `candidate_index` fields into vectors (nested in the case of responses with `candidate_index`).

In order to support multiple candidate addresses in the US Street API, each address in the input will correspond to a vector of zero or more candidate addresses.

Other changes:

* We now use the `democracyworks` groupId to be a bit more neighborly
* Version is tentatively bumped to `0.2.0-SNAPSHOT` - if released, it will be released at `0.2.0`

gws committed:
Remove placeholder documentation

gws committed:
Update artifact name, version, and dependencies

This is in preparation to add a parallel set of namespaces with an
improved API.

gws committed:
Add a new API client record and low-level HTTP functionality

gws committed:
Add SmartyStreets' US Street Address API

gws committed:
Add SmartyStreets' US ZIP Code API

gws committed:
Update documentation to reflect the current functionality

gws committed:
Add a bare-bones dev REPL environment

zonotope commented on src/democracyworks/smartystreets/core.clj:
```diff
@@ -0,0 +1,25 @@
+  [auth-id auth-token]
+  (->Client auth-id auth-token))
+
+(defn http-get
```
perhaps `get` and `post` should be protocol methods on the client? it doesn't make a difference either way, but it might be nice to clarify the api.

gws replied:
I actually had `fetch-one` and `fetch-many` in the API namespaces defined as a protocol and extended `Client` to each of them, to clarify the API. Switching back to that would not be a breaking change, if that's desired.

Here, `http-get` and `http-post` are little more than convenience wrappers with some default options, so it's implementation detail (specific to Client) and not API.

zonotope reviewed (approved):
i have some *really* minor suggestions/questions, but this looks good to me.

gws commented on src/democracyworks/smartystreets/api/us_street.clj:
```diff
@@ -0,0 +1,57 @@
+
+  `request` should be an HTTP input request, described here:
+
+  https://smartystreets.com/docs/cloud/us-street-api#input-fields
```
Yeah, `name` gets called on keys, so it'll work with either the keyword or string versions, and it always returns keywords.

sduckett commented on src/democracyworks/smartystreets/api/us_street.clj:
```diff
@@ -0,0 +1,57 @@
+(def base-url
+  "https://us-street.api.smartystreets.com/street-address")
+
+(defn parse-response
```
For [more about `:input_index` and `:candidate_index`](https://smartystreets.com/docs/cloud/us-street-api#root).

sduckett commented on src/democracyworks/smartystreets/api/us_zipcode.clj:
```diff
@@ -0,0 +1,47 @@
+(def base-url
+  "https://us-zipcode.api.smartystreets.com/lookup")
+
+(defn parse-response
```
For [more about the response we're parsing](https://smartystreets.com/docs/cloud/us-zipcode-api#http-response-output).

sduckett reviewed (commented):
Here are some thoughts I had while reading the diff.

gws committed:
Remove unnecessary Travis-CI configuration

gws committed:
Bump dependencies

gws committed:
Add 3-arity Client record constructor

gws committed:
Update documentation

gws commented:
@sduckett @zonotope 

I've made a few changes (some incorporating your feedback):

- Updated all docstrings.
- We describe each SmartyStreets API using protocols to make it easier to test code that calls into the library.
- I flattened the namespace hierarchy a little.
- I've added a way to pass through options to `clj-http` in case clients need to set timeouts or do other things I haven't thought of, though the defaults should be sufficient for most use cases.

Assuming nothing objectionable about the API, I'm probably going to merge this and cut a release this week.

gws committed:
Update dependencies
